src/chart/generateCategoricalChart.tsx

```diff
@@ -288,87 +288,115 @@ const getAxisMapByAxes = (
 
     let domain, duplicateDomain, categoricalDomain;
 
-    if (dataKey) {
-      // has dataKey in <Axis />
-      domain = getDomainOfDataByKey(displayedData, dataKey, type);
-
-      if (type === 'category' && isCategorical) {
-        // the field type is category data and this axis is catrgorical axis
-        const duplicate = hasDuplicate(domain);
-
-        if (allowDuplicatedCategory && duplicate) {
-          duplicateDomain = domain;
-          // When category axis has duplicated text, serial numbers are used to generate scale
-          domain = _.range(0, len);
-        } else if (!allowDuplicatedCategory) {
-          // remove duplicated category
-          domain = parseDomainOfCategoryAxis(child.props.domain, domain, child).reduce(
-            (finalDomain: any, entry: any) => (finalDomain.indexOf(entry) >= 0 ? finalDomain : [...finalDomain, entry]),
-            [],
+    /*
+     * This is a hack to short-circuit the domain creation here to enhance performance.
+     * Usually, the data is used to determine the domain, but when the user specifies
+     * a domain upfront (via props), there is no need to calculate the domain start and end,
+     * which is very expensive for a larger amount of data.
+     * The only thing that would prohibit short-circuiting is when the user doesn't allow data overflow,
+     * because the axis is supposed to ignore the specified domain that way.
+     */
+    if (type === 'number' && allowDataOverflow === true) {
+      const specifiedDomain: unknown[] | null | undefined = child.props.domain;
+      const domainStart: unknown | null | undefined = specifiedDomain?.[0];
+      const domainEnd: unknown | null | undefined = specifiedDomain?.[1];
```

Comment:
nit: can we destructure this?
```suggestion
      const [domainStart, domainEnd]: [unknown | null | undefined, unknown | null | undefined] = specifiedDomain;
```
pretty sure JS doesn't throw errors even if the array is `[]` or `[200, undefined]`

Reply:
Just using destructuring does not work, as `specifiedDomain` can be `undefined` there.
I could add another check in the `if` if you want, but using but using optional chaining takes care of all cases I think:

![image](https://user-images.githubusercontent.com/29319414/215167737-d92be808-b349-4b57-8ecb-d97a656c5cc9.png)

Reply:
Yeah it does, idk what I was thinking 😅